File: 2023/day3/golang/solution.go
package main

import (
	"regexp"
	"strconv"
	"strings"
)

func partOne(lines []string) int {
	var results []int
	specialCharacters := "!@#$%^&*()-+?_=,<>/"

	for linesIdx, line := range lines {
		re := regexp.MustCompile(`\d+`)
		matches := re.FindAllStringIndex(line, -1)

		for _, match := range matches {
			x1 := max(match[0]-1, 0)
			x2 := min(match[1]+1, len(line))

			y1 := max(linesIdx-1, 0)
			y2 := min(linesIdx+1, len(lines)-1)

			checks := line[x1:x2]

			if y1 != linesIdx {
				checks += lines[y1][x1:x2]
			}

			if y2 != linesIdx {
				checks += lines[y2][x1:x2]
			}

			for _, char := range specialCharacters {
				if strings.Contains(checks, string(char)) {
					matchedVal, _ := strconv.Atoi(line[match[0]:match[1]])
					results = append(results, matchedVal)
				}
			}
		}
	}

	sum := 0
	for _, val := range results {
		sum += val
	}

	return sum
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func partTwo(lines []string) int {
	var results [][]string
	digitRegex := regexp.MustCompile(`\d+`)
	starRegex := regexp.MustCompile(`\*`)

	for linesIdx, line := range lines {
		for _, match := range digitRegex.FindAllStringIndex(line, -1) {
			x1 := max(match[0]-1, 0)
			x2 := min(match[1]+1, len(line))

			y1 := max(linesIdx-1, 0)
			y2 := min(linesIdx+1, len(lines)-1)

			checkAndAppend := func(y, x1, x2 int, matchValue string) {
				if starRegex.MatchString(lines[y][x1:x2]) {
					starIdx := -1
					for _, m := range starRegex.FindAllStringIndex(lines[y], -1) {
						if x1 <= m[0] && x2 >= m[1] {
							starIdx = m[0]
							break
						}
					}
					if starIdx != -1 {
						results = append(results, []string{strconv.Itoa(y) + strconv.Itoa(starIdx), matchValue})
					}
				}
			}

			matchValue := line[match[0]:match[1]]
			checkAndAppend(linesIdx, x1, x2, matchValue)

			if y1 != linesIdx {
				checkAndAppend(y1, x1, x2, matchValue)
			}

			if y2 != linesIdx {
				checkAndAppend(y2, x1, x2, matchValue)
			}
		}
	}

	dict := make(map[string][]int)
	for _, r := range results {
		key := r[0]
		value, _ := strconv.Atoi(r[1])
		dict[key] = append(dict[key], value)
	}

	res := 0
	for _, value := range dict {
		if len(value) == 2 {
			res += value[0] * value[1]
		}
	}

	return res
}

File: 2023/day3/solution.py
from typing import List
import os
import re

special_characters = "!@#$%^&*()-+?_=,<>/"


def part1(lines: List[str]):
    results = []

    for lines_idx, line in enumerate(lines):
        for match in re.finditer(r"\d+", line.strip()):
            x1 = match.start() - 1 if match.start() > 0 else 0
            x2 = match.end() + 1 if match.end() < len(line) - 1 else len(line) - 1

            y1 = lines_idx - 1 if lines_idx > 0 else 0
            y2 = lines_idx + 1 if lines_idx < len(lines) - 1 else len(lines) - 1

            checks = line[x1:x2]

            if y1 is not lines_idx:
                checks += lines[y1][x1:x2]

            if y2 is not lines_idx:
                checks += lines[y2][x1:x2]

            results.extend(match.group() for i in special_characters if i in checks)
    results = list(map(lambda x: int(x), results))
    return sum(results)


def part2(lines: List[str]):
    results = []

    for lines_idx, line in enumerate(lines):
        for match in re.finditer(r"\d+", line.strip()):
            x1 = match.start() - 1 if match.start() > 0 else 0
            x2 = match.end() + 1 if match.end() < len(line) - 1 else len(line) - 1

            y1 = lines_idx - 1 if lines_idx > 0 else 0
            y2 = lines_idx + 1 if lines_idx < len(lines) - 1 else len(lines) - 1

            if "*" in line[x1:x2]:
                star_idx = -1
                for m in re.finditer(r"\*", line):
                    if x1 <= m.start() and x2 >= m.end():
                        star_idx = m.start()
                results.append([str(lines_idx) + str(star_idx), int(match.group())])

            if y1 is not lines_idx and "*" in lines[y1][x1:x2]:
                star_idx = -1
                for m in re.finditer(r"\*", lines[y1]):
                    if x1 <= m.start() and x2 >= m.end():
                        star_idx = m.start()
                results.append([str(y1) + str(star_idx), int(match.group())])

            if y2 is not lines_idx and "*" in lines[y2][x1:x2]:
                star_idx = -1
                for m in re.finditer(r"\*", lines[y2]):
                    if x1 <= m.start() and x2 >= m.end():
                        star_idx = m.start()
                results.append([str(y2) + str(star_idx), int(match.group())])

    d = {}
    for r in results:
        if r[0] in d:
            d[r[0]].append(r[1])
        else:
            d[r[0]] = [r[1]]

    return sum(value[0] * value[1] for value in d.values() if len(value) == 2)
